### PR TITLE
Move capture_aes_random_batch function to capture.py.

### PR DESCRIPTION
**Pull request by abdullahvarici**
Rename simple_capture_traces.py to capture.py. Delete
simple_capture_traces_batch.py.

Signed-off-by: Abdullah Varici <abdullah.varici@lowrisc.org>

### CODE REVIEW

**Comment by abdullahvarici** [comment]
I have started with the aes_random_batch function. Instead of using scared library I've used crypto.cipher for AES encryption. Waverunner is not supported but if it should be supported, I think we can change the code to support for all capture functions in a generic way. I also have added trace range check.

File: cw/cw305/capture.py
Comment (vogelpi) [inline cw/cw305/capture.py:343]: This should go into a seperate commit.

File: cw/cw305/capture.py
Comment (vogelpi) [inline cw/cw305/capture.py:348]: These changes are unrelated to the batch capture work. They should go into a seperate commit. Also, it introduces again lint errors. Can you please run `flake8 capture.py --max-line-length 100` and fix all errors. Thanks!

File: cw/cw305/capture.py
Comment (vogelpi) [inline cw/cw305/capture.py:357]: Nit: This causes lint errors.

File: cw/cw305/capture.py
Comment (vogelpi) [inline cw/cw305/capture.py:368]: Same here: lint + it's not related to the batch capture change.

File: doc/getting_started.md
Comment (vogelpi) [inline doc/getting_started.md:225]: Please revert this line as it assumes the FPGA has already been loaded previously. Otherwise it would just print:
```
Connecting and loading FPGA... Done!
```
as before.

File: doc/getting_started.md
Comment (vogelpi) [inline doc/getting_started.md:228]: Please remove these lines and update your scope firmware as described under https://chipwhisperer.readthedocs.io/en/latest/api.html#firmware-update

File: doc/getting_started.md
Comment (vogelpi) [inline doc/getting_started.md:249]: maybe you can just replace `abdullah` here with `user` to make the guide more generic.

File: doc/getting_started.md
Comment (vogelpi) [inline doc/getting_started.md:248]: I think this assumes that you've changed the `num_traces` variable in the capture_aes.yaml file. Do you want to update the command above accordingly?

File: doc/getting_started.md
Comment (vogelpi) [inline doc/getting_started.md:467]: `When using a CW-Husky board, the script also supports capturing AES power traces in batches to achieve substantially higher capture rates (ca. 1300 traces/s).`

File: doc/getting_started.md
Comment (vogelpi) [inline doc/getting_started.md:508]: I would select a higher number than 100 here. Say 5000.

File: doc/getting_started.md
Comment (vogelpi) [inline doc/getting_started.md:520]: Both these points are no longer required. The ADC outputs are now checked within the capture loop and independent of the plot setting. And thanks to your work, one can just provide a different `num_traces` argument using `--num-traces 2000000` :-)

File: doc/getting_started.md
Comment (vogelpi) [inline doc/getting_started.md:578]: In this section can you please also remove steps 3 (set capture.num_traces) and 4 (set plot_capture.show) and instead adjust the capture command below?

File: doc/getting_started.md
Comment (vogelpi) [inline doc/getting_started.md:613]: Here we would need an argument `--num-traces 2000000`

**Review by vogelpi — COMMENTED** [review]
Thanks @abdullahvarici for working on this! This looks already pretty good and it will simplify a lot future work using batch capture. This is cool :-)

I've added a couple of comments. It would be great if this PR could keep the focus on the batch capture. Moving the mix columns capture can go into a separate PR to simplify review. Also it would be great if we could keep the WaveRunner support. We can't test it locally but maybe @alphan will find time to do this. If not, we at least keep around the code and it will be easier in the future to re-enable WaveRunner support.

Last but not least, I would just keep using `scared` to compute the plaintexts/ciphertexts in parallel instead of switchting to the other implementation. Or this there a particular reason for doing this?

File: cw/cw305/capture.py
Comment (abdullahvarici) [reply cw/cw305/capture.py:215]: In other functions in the same script we are using crypto.cipher and I thought it would be better if we use the same structure. In addition scared library is slower than crypto.cipher. When we try to capture 100k traces with scared we get 1697 traces/s and with crypto.cipher we get 1735 traces/s.

File: cw/cw305/capture.py
Comment (abdullahvarici) [reply cw/cw305/capture.py:244]: I copied to there by mistake I think. I take it back.

File: cw/cw305/capture.py
Comment (abdullahvarici) [reply cw/cw305/capture.py:348]: Oh OK.

File: cw/cw305/capture.py
Comment (alphan) [inline cw/cw305/capture.py:32]: nit: enums are safer:

```suggestion
from enum import Enum

...

class ScopeType(str, Enum):
    cw = "cw"
    waverunner = "waverunner"

opt_scope_type = typer.Option(ScopeType.cw, help=("Scope type"))
```

Please take a look at [here](https://typer.tiangolo.com/tutorial/parameter-types/enum/) for more details.

File: doc/getting_started.md
Comment (alphan) [inline doc/getting_started.md:458]: nit: Since we have a single script, consider removing the list and rewording this part.

File: cw/cw305/capture.py
Comment (alphan) [inline cw/cw305/capture.py:227]: let's keep using `scared.aes.base.encrypt()` here.

File: cw/cw305/capture.py
Comment (alphan) [inline cw/cw305/capture.py:259]: nit: can't remember if this was a recent change but this call is missing `dtype=np.uint16`. pointing out to avoid surprises down the road.

File: cw/cw305/capture.py
Comment (alphan) [inline cw/cw305/capture.py:194]:
```suggestion
```

File: cw/cw305/capture.py
Comment (abdullahvarici) [reply cw/cw305/capture.py:193]: For checking, I moved `random.seed()` before `capture_init()` and saw that the same sequence of key and pt is generated.

File: cw/cw305/capture.py
Comment (abdullahvarici) [reply cw/cw305/capture.py:227]: Is there a specific reason for that? In other functions in the same script we are using crypto.cipher and I thought it would be better if we use the same structure. In addition scared library is slower than crypto.cipher. When we try to capture 100k traces with scared we get 1697 traces/s and with crypto.cipher we get 1735 traces/s.

File: cw/cw305/capture.py
Comment (abdullahvarici) [reply cw/cw305/capture.py:259]: When I try to add `np.uint16` as an argument to `cw.common.traces.Trace()` function, it gives an error saying "takes 5 positional arguments but 6 were given".

File: cw/cw305/capture.py
Comment (abdullahvarici) [reply cw/cw305/capture.py:194]: Should we use scared function instead?

**Comment by abdullahvarici** [comment]
Thanks @alphan for the review. I have already applied and saw that different capture functions works with different parameters. I have made changes according to your review and forced-pushed them.

File: cw/cw305/capture.py
Comment (vogelpi) [reply cw/cw305/capture.py]: Thanks.

File: cw/cw305/capture.py
Comment (vogelpi) [reply cw/cw305/capture.py]: Thanks.

File: doc/getting_started.md
Comment (vogelpi) [reply doc/getting_started.md:556]: Thanks.

File: doc/getting_started.md
Comment (vogelpi) [reply doc/getting_started.md]: Actually, you now wrote `--num-traces=2000000` does that work too or do we need `--num-traces 2000000` (without equal sign)?

File: cw/cw305/capture.py
Comment (vogelpi) [reply cw/cw305/capture.py:233]: I don't remember to be honest. But in general I would just change one thing at a time (merge batch script into normal script for this PR). The difference in performance you see is roughly 2%. It's not too much but maybe also not negligible. This change could be another PR. Up to now, two people have reviewed this PR and this point has been brought up by both. So please move the change into a separate PR to unblock this one.

File: cw/cw305/capture.py
Comment (vogelpi) [reply cw/cw305/capture.py:266]: The `dtype=np.uint16` argument is required for the `append()` function not the `Trace()`. Without this argument, the trace files will again be 4x bigger. This has been added very recently and it probably got forgotten when you rebased your work. 

This is a good example why we don't want to change more than one thing at once/per PR. Please read this section from the contribution guide to learn how we structure PRs and commits: https://docs.opentitan.org/doc/project/detailed_contribution_guide/#working-with-pull-requests

File: cw/cw305/capture.py
Comment (vogelpi) [reply cw/cw305/capture.py:200]: No, let's not do that now. We can file a separate PR to use the same cipher everywhere. This PR should just be about merging the batch script into the regular one.

File: cw/cw305/capture.py
Comment (vogelpi) [reply cw/cw305/capture.py:266]: Good catch @alphan by the way!

File: doc/getting_started.md
Comment (abdullahvarici) [reply doc/getting_started.md]: My mistake :) Changed it.

**Review by vogelpi — APPROVED** [review]
Thanks a lot @abdullahvarici for implementing the remaining feedback! Maybe we wait for a couple more hours before merging in case @alphan want to take another look. But from my side, this looks definitely good!

File: cw/cw305/capture.py
Comment (vogelpi) [reply cw/cw305/capture.py:270]: Cool, I've tried out the PR. 100000 traces are ~235 MiB in size. So everything as expected ;-)

**Comment by alphan** [comment]
> Thanks a lot @abdullahvarici for implementing the remaining feedback! Maybe we wait for a couple more hours before merging in case @alphan want to take another look. But from my side, this looks definitely good!

Apologies for the delay, LGTM, thanks @abdullahvarici!

**Pull request merged**
